doc/developer/design/20210406_mapfilterproject.md

```diff
@@ -0,0 +1,155 @@
+# MapFilterProject Pushdown
+
+## Summary
+
+<!--
+// Brief, high-level overview. A few sentences long.
+// Be sure to capture the customer impact - framing this as a release note may be useful.
+-->
+
+Dataflow computations consume data from sources, but rarely require all of the information that a source can provide.
+Some simple transformation, filtering, and projection can be applied in-place on the data, as (or even before) it arrives to the dataflow.
+This proposal presents one way of wrapping and communicating this information to sources, who can then implement it if appropriate.
+
+## Goals
+
+<!--
+// Enumerate the concrete goals that are in scope for the project.
+-->
+
+The main goal is to define a concrete encapsulation of operations a source can apply to its stream of records before presenting them to dataflow.
+In addition, we want to assess the fitness for use of this encapsulation, to determine that it is (or is not) helpful for sources.
+We intend to frame an interface with sources that allow them to communicate back to dataflow what of the operations they were able to perform.
+
+## Non-Goals
+
+<!--
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+We do not intend to determine the policy for which operators should be communicated to sources and which should be held back.
+We expect a substantial amount of follow-up work on source implementations where we determine how best to apply these operators.
+We could go further with the set of operations that we push in to sources, including e.g. `FlatMap`, `Negate`, or `Union` operators.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+We chose the name `MapFilterProject` to represent the idea that we are capturing a sequence of `Map`, then `Filter`, then `Project` statements.
+
+The `MirRelationExpr::Map` operator contains a sequence of `MirScalarExpr` expressions.
+Each expression is a function that is applied to the existing columns, and which produces a new column.
+The sequence of expressions indicates a sequence of new columns to add to each row input to the `Map`.
+
+The `MirRelationExpr::Filter` operator contains a sequence of `MirScalarExpr` expressions.
+Each expression is a function that is applied to the existing columns, and which produces either `Datum::True` or not.
+Rows that produce `Datum::True` are retained in the output, and rows that do not are discarded.
```

Comment:
I think this is implicit, but rows that produce true for all expressions are retained in the output right?

Reply:
Ah, do you mean that it isn't clearly stated that it is AND rather than OR (or something else)? Happy to add that with a clarifying comment for @wangandi's comment too.

Reply:
yup that its an AND